legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabEntityApi.java

```diff
@@ -917,9 +917,41 @@ synchronized Entity getEntity()
                 Entity localEntity = this.sourceDirectory.deserialize(this.file);
                 if (!Objects.equals(localEntity.getPath(), getEntityPath()))
                 {
-                    throw new RuntimeException("Expected entity path " + getEntityPath() + ", found " + localEntity.getPath());
+                    /**
+                     *  In gitlab, each entity will be stored as a file of which file name is built upon entity's path.
+                     *  for now, the file name of function is functionName, which doesn't contain function signature.
+                     *  To support function overloading, Engine is updated to store function signature as the name of
+                     *  PackageableElement, which updates PackageableElement's path at the same time since path is
+                     *  built upon PackageableElement's name. Any new function entity will be stored in a file whose
+                     *  name is the full function signature in gitlab. However, existing functions need to be updated
+                     *  to be consistent.
+                     *
+                     *  getEntityPath() is from file name in gitlab
+                     *  localEntity is computed from Engine, once Engine is updated, it will contain the whole function signature.
+                     *  For existing functions in gitlab, localEntity.getPath() and getEntityPath() won't match anymore.
+                     *  However, they are related. localEntity.getPath() must contain getEntityPath() if everything is correct
+                     *  for these cases. If condition below is true, a new entity will be created using the short version
+                     *  of the path (functionName only) as a flag for Studio to trigger a `delete` and `create` actions.
+                     *  Users have to push local changes computed automatically for functions above, which will delete old files
+                     *  and create corresponding new files for functions. Therefore, gitlab will keep clean and consistent.
+                     */
+                    if ((localEntity.getPath() != null) && (getEntityPath() != null) &&
+                            localEntity.getPath().contains(getEntityPath()) &&
+                            "meta::pure::metamodel::function::ConcreteFunctionDefinition".equals(localEntity.getClassifierPath()))
+                    {
+                        // getEntityPath() only contains functionName.
+                        this.entity = Entity.newEntity(getEntityPath(), localEntity.getClassifierPath(), localEntity.getContent());
```

Comment:
Won't this mean that there will be an inconsistency between the entity path and the name property in the content?

Reply:
Yes :) , but they will be consistent again once users push automatically computed local changes for functions. 
entities before push:
```
[
  {
    "classifierPath": "meta::pure::metamodel::function::ConcreteFunctionDefinition",
    "path": "model::NewEnumeration",
    "content": {
      "_type": "function",
      "name": "NewEnumeration_String_$1345_19008$__String_$0_1$_",
      "preConstraints": [],
      "postConstraints": [],
      "parameters": [
        {
          "_type": "var",
          "name": "param1",
          "multiplicity": {
            "lowerBound": 1345,
            "upperBound": 19008
          },
          "class": "String"
        }
      ],
      "returnType": "String",
      "returnMultiplicity": {
        "lowerBound": 0,
        "upperBound": 1
      },
      "stereotypes": [],
      "taggedValues": [],
      "body": [
        {
          "_type": "string",
          "values": [
            ""
          ],
          "multiplicity": {
            "lowerBound": 1,
            "upperBound": 1
          }
        }
      ],
      "package": "model"
    }
  },
  {
    "classifierPath": "meta::pure::metamodel::function::ConcreteFunctionDefinition",
    "path": "model::NewEnumeration_String_$1_MANY$__NewClass_$0_1$__String_1_",
    "content": {
      "_type": "function",
      "name": "NewEnumeration_String_$1_MANY$__NewClass_$0_1$__String_1_",
      "preConstraints": [],
      "postConstraints": [],
      "parameters": [
        {
          "_type": "var",
          "name": "param1",
          "multiplicity": {
            "lowerBound": 1
          },
          "class": "String"
        },
        {
          "_type": "var",
          "name": "pa",
          "multiplicity": {
            "lowerBound": 0,
            "upperBound": 1
          },
          "class": "model::NewClass"
        }
      ],
      "returnType": "String",
      "returnMultiplicity": {
        "lowerBound": 1,
        "upperBound": 1
      },
      "stereotypes": [],
      "taggedValues": [],
      "body": [
        {
          "_type": "string",
          "values": [
            ""
          ],
          "multiplicity": {
            "lowerBound": 1,
            "upperBound": 1
          }
        }
      ],
      "package": "model"
    }
  }
]
```
push actions: 
<img width="596" alt="Screen Shot 2022-09-26 at 2 08 14 PM" src="https://user-images.githubusercontent.com/73408381/192349133-cf6c6728-4b88-469a-bb6f-563531127202.png">


entities after push:

```
[
  {
    "classifierPath": "meta::pure::metamodel::function::ConcreteFunctionDefinition",
    "path": "model::NewEnumeration_String_$1_MANY$__NewClass_$0_1$__String_1_",
    "content": {
      "_type": "function",
      "name": "NewEnumeration_String_$1_MANY$__NewClass_$0_1$__String_1_",
      "preConstraints": [],
      "postConstraints": [],
      "parameters": [
        {
          "_type": "var",
          "name": "param1",
          "multiplicity": {
            "lowerBound": 1
          },
          "class": "String"
        },
        {
          "_type": "var",
          "name": "pa",
          "multiplicity": {
            "lowerBound": 0,
            "upperBound": 1
          },
          "class": "model::NewClass"
        }
      ],
      "returnType": "String",
      "returnMultiplicity": {
        "lowerBound": 1,
        "upperBound": 1
      },
      "stereotypes": [],
      "taggedValues": [],
      "body": [
        {
          "_type": "string",
          "values": [
            ""
          ],
          "multiplicity": {
            "lowerBound": 1,
            "upperBound": 1
          }
        }
      ],
      "package": "model"
    }
  },
  {
    "classifierPath": "meta::pure::metamodel::type::Class",
    "path": "model::NewClass",
    "content": {
      "_type": "class",
      "name": "NewClass",
      "superTypes": [],
      "originalMilestonedProperties": [],
      "properties": [],
      "qualifiedProperties": [],
      "stereotypes": [],
      "taggedValues": [],
      "constraints": [],
      "package": "model"
    }
  }
]
```

Reply:
It's not ok for the entity API to return something that's inconsistent. It doesn't matter whether it will be corrected by some subsequent user action. The entity path _must_ equal package + "::" + name. So if you're going to change the entity path, you must also change the content to match.

Reply:
Thanks for your review!  I updated the code to obey the rule.
The entities before push: 

```
[
  {
    "classifierPath": "meta::pure::metamodel::function::ConcreteFunctionDefinition",
    "path": "model::NewEnumeration",
    "content": {
      "stereotypes": [],
      "taggedValues": [],
      "package": "model",
      "_type": "function",
      "name": "NewEnumeration",
      "returnMultiplicity": {
        "lowerBound": 1,
        "upperBound": 1
      },
      "body": [
        {
          "_type": "string",
          "values": [
            ""
          ],
          "multiplicity": {
            "lowerBound": 1,
            "upperBound": 1
          }
        }
      ],
      "parameters": [
        {
          "_type": "var",
          "name": "param1",
          "multiplicity": {
            "lowerBound": 1
          },
          "class": "String"
        },
        {
          "_type": "var",
          "name": "pa",
          "multiplicity": {
            "lowerBound": 0,
            "upperBound": 1
          },
          "class": "model::NewClass"
        }
      ],
      "returnType": "String",
      "preConstraints": [],
      "postConstraints": []
    }
  },
  {
    "classifierPath": "meta::pure::metamodel::function::ConcreteFunctionDefinition",
    "path": "model::NewEnumeration_String_$1345_19008$__String_$0_1$_",
    "content": {
      "_type": "function",
      "name": "NewEnumeration_String_$1345_19008$__String_$0_1$_",
      "preConstraints": [],
      "postConstraints": [],
      "parameters": [
        {
          "_type": "var",
          "name": "param1",
          "multiplicity": {
            "lowerBound": 1345,
            "upperBound": 19008
          },
          "class": "String"
        }
      ],
      "returnType": "String",
      "returnMultiplicity": {
        "lowerBound": 0,
        "upperBound": 1
      },
      "stereotypes": [],
      "taggedValues": [],
      "body": [
        {
          "_type": "string",
          "values": [
            ""
          ],
          "multiplicity": {
            "lowerBound": 1,
            "upperBound": 1
          }
        }
      ],
      "package": "model"
    }
  }
]
```